### PR TITLE
docs: Add ERC20 proxy contract detection specs

### PR DESCRIPTION
**Pull request by lrnt**
## Description of the changes

Added a draft of the new payment network using a proxy contract for ERC20 payments.

To be discussed: both ETH (with input data) and ERC20 (with proxy contract) payment detection will use a salted hash to link the payment with the request. It would be interesting if we came up with a name for this salted hash since it serves the same purpose. I suggest using payment reference (`paymentReference`). Other suggestions are more than welcome.

## Link to Jira

https://requestnetwork.atlassian.net/browse/PROT-1013

### CODE REVIEW

File: packages/advanced-logic/specs/payment-network-erc20-proxy-contract-0.1.0-DRAFT.md [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+- you are curious enough to dive and see what is under the hood of the Request protocol
+
+Prerequisite: Having read the advanced logic specification (see [here](./advanced-logic-specs-0.1.0-DRAFT.md)).
+
+## Description
+
+This extension allows the payments and the refunds to be made in ERC20 tokens on the Ethereum blockchain.
+The payment is made through a proxy contract. This proxy contract does the ERC20 token transfer on behalf of
+the user. The contract ensures a link between an ERC20 transfer and a request through a `paymentReference`.
+This `paymentReference` conists of the last 8 bytes of a salted hash of the requestId: `last8Bytes(hash(requestId + salt + address))`:
```
Comment (benjlevesque) [inline packages/advanced-logic/specs/payment-network-erc20-proxy-contract-0.1.0-DRAFT.md]:
```suggestion
This `paymentReference` consists of the last 8 bytes of a salted hash of the requestId: `last8Bytes(hash(requestId + salt + address))`:
```

File: packages/advanced-logic/specs/payment-network-erc20-proxy-contract-0.1.0-DRAFT.md [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+
+```
+pragma solidity 0.5.11;
+
+import "openzeppelin-solidity/contracts/token/ERC20/ERC20.sol";
+
+contract RequestErc20Proxy {
+    event RequestTransfer(address tokenAddress, address indexed to, uint256 amount, bytes indexed requestData);
+
+    function transferFromWithData(
```
Comment (benjlevesque) [inline packages/advanced-logic/specs/payment-network-erc20-proxy-contract-0.1.0-DRAFT.md]: earlier (L28, L35) this method is referenced as `transferWithData`, typo?

File: packages/advanced-logic/specs/payment-network-erc20-proxy-contract-0.1.0-DRAFT.md [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+The contract contains one function called `transferWithData` which takes 4 arguments:
+
+- `tokenAddress` is the address of the ERC20 contract
+- `to` is the destination address for the tokens
+- `amount` is the amount of tokens
+- `requestData` is the reference data used to track the transfer (see `paymentReference`)
+
+The `RequestTransfer` event is emitted when the tokens are transfered. This event contains the same 4 arguments as the `transferWithData` function.
+
+```
```
Comment (benjlevesque) [inline packages/advanced-logic/specs/payment-network-erc20-proxy-contract-0.1.0-DRAFT.md]:
```suggestion
```solidity
```

File: packages/advanced-logic/specs/payment-network-erc20-proxy-contract-0.1.0-DRAFT.md [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+| Warning                                 | Condition                                                   |
+| --------------------------------------- | ----------------------------------------------------------- |
+| "paymentAddress is given by the payer"  | If `signer` is the payer **and** `paymentAddress` is given  |
+| "refundAddress is given by the payee"   | If `signer` is the payee **and** `refundAddress` is given   |
+
+Note: These warnings are necessary to highlight to avoid attempts of fake payments and refunds. For example, a payer could create a request using as the payment address one of his own addresses. A system could interpret a transaction to this address as a payment while the payee did not receive the funds.
+
+#### Results
+
+A extension state is created with the following properties:
```
Comment (benjlevesque) [inline packages/advanced-logic/specs/payment-network-erc20-proxy-contract-0.1.0-DRAFT.md]:
```suggestion
An extension state is created with the following properties:
```

File: packages/advanced-logic/specs/payment-network-erc20-proxy-contract-0.1.0-DRAFT.md [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+- The signer is the `payee`
+- The extension property `paymentAddress` is undefined
+
+##### Warnings
+
+None.
+
+##### Results
+
+A extension state is updated with the following properties:
```
Comment (benjlevesque) [inline packages/advanced-logic/specs/payment-network-erc20-proxy-contract-0.1.0-DRAFT.md]:
```suggestion
An extension state is updated with the following properties:
```

File: packages/advanced-logic/specs/payment-network-erc20-proxy-contract-0.1.0-DRAFT.md [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+- The signer is the `payer`
+- The extension property `refundAddress` is undefined
+
+##### Warnings
+
+None.
+
+##### Results
+
+A extension state is updated with the following properties:
```
Comment (benjlevesque) [inline packages/advanced-logic/specs/payment-network-erc20-proxy-contract-0.1.0-DRAFT.md]:
```suggestion
An extension state is updated with the following properties:
```

File: packages/advanced-logic/specs/payment-network-erc20-proxy-contract-0.1.0-DRAFT.md [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+| Property                  | Type   | Description                                    | Requirement   |
+| ------------------------- | ------ | ---------------------------------------------- | ------------- |
+| **id**                    | String | constant value: "pn-erc20-proxy-contract"      | **Mandatory** |
+| **type**                  | String | constant value: "paymentNetwork"               | **Mandatory** |
+| **version**               | String | constant value: "0.1.0"                        | **Mandatory** |
+| **events**                | Array  | List of the actions performed by the extension | **Mandatory** |
+| **values**                | Object |                                                |               |
+| **values.paymentAddress** | String | Ethereum address for the payment               | Optional      |
+| **values.refundAddress**  | String | Ethereum address for the refund                | Optional      |
+| **values.salt**           | String | Salt for the request                           | Optional      |
```
Comment (vrolland) [inline packages/advanced-logic/specs/payment-network-erc20-proxy-contract-0.1.0-DRAFT.md]: The salt is **Mandatory**

File: packages/advanced-logic/specs/payment-network-erc20-proxy-contract-0.1.0-DRAFT.md [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+---
+
+## Interpretation
+
+The proxy contract address is determined by the `request.currency.network` (see (table)[#Contract] with proxy contract addresses).
+
+Any `RequestTransfer` events emitted from the proxy contract with the following arguments are considered as a payment:
+- `tokenAddress` `===` `request.currency.value`
+- `to` `===` `paymentAddress`
+- `requestData` `===` `last8Bytes(hash(requestId + salt + payment address))`
```
Comment (vrolland) [inline packages/advanced-logic/specs/payment-network-erc20-proxy-contract-0.1.0-DRAFT.md]: `last8Bytes(hash(lowercase(requestId + salt + payment address)))`

File: packages/advanced-logic/specs/payment-network-erc20-proxy-contract-0.1.0-DRAFT.md [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+
+Any `RequestTransfer` events emitted from the proxy contract with the following arguments are considered as a refund:
+- `tokenAddress` `===` `request.currency.value`
+- `to` `===` `refundAddress`
+- `requestData` `===` `last8Bytes(hash(requestId + salt + payment address))`
```
Comment (vrolland) [inline packages/advanced-logic/specs/payment-network-erc20-proxy-contract-0.1.0-DRAFT.md]: `last8Bytes(hash(lowercase(requestId + salt + refund address)))`

Comment (vrolland) [inline packages/advanced-logic/specs/payment-network-erc20-proxy-contract-0.1.0-DRAFT.md]: payment address => refund address

File: packages/advanced-logic/specs/payment-network-erc20-proxy-contract-0.1.0-DRAFT.md [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+None.
+
+##### Results
+
+An extension state is updated with the following properties:
+
+|  Property                  |  Value                                               |
+| -------------------------- | ---------------------------------------------------- |
+| **values.paymentAddress**  | `paymentAddress` from parameters                     |
+| **parameters.paymentSalt** | Salt for the payment                                 |
```
Comment (vrolland) [inline packages/advanced-logic/specs/payment-network-erc20-proxy-contract-0.1.0-DRAFT.md]: `| **parameters.paymentSalt** | Salt for the payment                                 |`
line to delete

File: packages/advanced-logic/specs/payment-network-erc20-proxy-contract-0.1.0-DRAFT.md [hunk 11]
```diff
@@ -0,0 +1,241 @@
...
+None.
+
+##### Results
+
+An extension state is updated with the following properties:
+
+|  Property                |  Value                                                 |
+| ------------------------ | ------------------------------------------------------ |
+| **values.refundAddress** | `refundAddress` from parameters                        |
+| **values.refundSalt**    | Salt for the refund                                    |
```
Comment (vrolland) [inline packages/advanced-logic/specs/payment-network-erc20-proxy-contract-0.1.0-DRAFT.md]: | **parameters.refundSalt** | Salt for the refund |
line to delete

File: packages/advanced-logic/specs/payment-network-erc20-proxy-contract-0.1.0-DRAFT.md [hunk 12]
```diff
@@ -0,0 +1,241 @@
...
+- you are curious enough to dive and see what is under the hood of the Request protocol
+
+Prerequisite: Having read the advanced logic specification (see [here](./advanced-logic-specs-0.1.0-DRAFT.md)).
+
+## Description
+
+This extension allows the payments and the refunds to be made in ERC20 tokens on the Ethereum blockchain.
+The payment is made through a proxy contract. This proxy contract does the ERC20 token transfer on behalf of
+the user. The contract ensures a link between an ERC20 transfer and a request through a `paymentReference`.
+This `paymentReference` consists of the last 8 bytes of a salted hash of the requestId: `last8Bytes(hash(requestId + salt + address))`:
```
[the remark this answers is not in the document]
Comment (romaric-juniet) [reply packages/advanced-logic/specs/payment-network-erc20-proxy-contract-0.1.0-DRAFT.md]: I changed it this week

File: packages/advanced-logic/specs/payment-network-erc20-proxy-contract-0.1.0-DRAFT.md [hunk 13]
```diff
@@ -0,0 +1,241 @@
...
+- `address` is the payment address for payments, the refund address for refunds
+- `hash()` is a keccak256 hash function
+- `last8Bytes()` take the last 8 bytes
+
+As a payment network, this extension allows to deduce a payment `balance` for the request. (see
+[Interpretation](#Interpretation))
+
+## Contract
+
+The contract contains one function called `transferWithData` which takes 4 arguments:
```
Comment (ltacker) [inline packages/advanced-logic/specs/payment-network-erc20-proxy-contract-0.1.0-DRAFT.md]: I put `transferFromWithReference` in depend if we consider the proxy smart contract could be used for something else

File: packages/advanced-logic/specs/payment-network-erc20-proxy-contract-0.1.0-DRAFT.md [hunk 14]
```diff
@@ -0,0 +1,241 @@
...
+
+- `tokenAddress` is the address of the ERC20 contract
+- `to` is the destination address for the tokens
+- `amount` is the amount of tokens
+- `requestData` is the reference data used to track the transfer (see `paymentReference`)
```
Comment (ltacker) [inline packages/advanced-logic/specs/payment-network-erc20-proxy-contract-0.1.0-DRAFT.md]: I would say `paymentReference`

File: packages/advanced-logic/specs/payment-network-erc20-proxy-contract-0.1.0-DRAFT.md [hunk 15]
```diff
@@ -0,0 +1,241 @@
...
+
+The `RequestTransfer` event is emitted when the tokens are transfered. This event contains the same 4 arguments as the `transferWithData` function.
+
+```solidity
+pragma solidity 0.5.11;
+
+import "openzeppelin-solidity/contracts/token/ERC20/ERC20.sol";
+
+contract RequestErc20Proxy {
+    event RequestTransfer(address tokenAddress, address indexed to, uint256 amount, bytes indexed requestData);
```
Comment (ltacker) [inline packages/advanced-logic/specs/payment-network-erc20-proxy-contract-0.1.0-DRAFT.md]: Good point from Romaric on another PR: the event name should be past participle.
I would name it `transferredWithReference`

File: packages/advanced-logic/specs/payment-network-erc20-proxy-contract-0.1.0-DRAFT.md [hunk 16]
```diff
@@ -0,0 +1,241 @@
...
+        ERC20 erc20 = ERC20(tokenAddress);
+        require(erc20.transferFrom(msg.sender, to, amount), 'transferFrom() failed');
+        emit RequestTransfer(tokenAddress, to, amount, requestData);
+    }
+}
+```
+
+| Network | Contract Address                           |
+|---------|--------------------------------------------|
+| Mainnet | TODO                                       |
```
[the remark this answers is not in the document]
Comment (ltacker) [reply packages/advanced-logic/specs/payment-network-erc20-proxy-contract-0.1.0-DRAFT.md]: No need, I added "Update erc20 proxy contract spec" to the mainnet deployment JIRA task

File: packages/advanced-logic/specs/payment-network-erc20-proxy-contract-0.1.0-DRAFT.md (hunk 15 above)
Comment (ltacker) [reply packages/advanced-logic/specs/payment-network-erc20-proxy-contract-0.1.0-DRAFT.md]: EDIT: not the case in fact

File: packages/advanced-logic/specs/payment-network-erc20-proxy-contract-0.1.0-DRAFT.md [hunk 17]
```diff
@@ -0,0 +1,223 @@
...
+| **type**                  | String | constant value: "paymentNetwork"               | **Mandatory** |
+| **version**               | String | constant value: "0.1.0"                        | **Mandatory** |
+| **events**                | Array  | List of the actions performed by the extension | **Mandatory** |
+| **values**                | Object |                                                |               |
+| **values.salt**           | String | Salt for the request                           | **Mandatory**      |
+| **values.paymentAddress** | String | Ethereum address for the payment               | Optional      |
+| **values.refundAddress**  | String | Ethereum address for the refund                | Optional      |
+
+
+Note: to use the Rinkeby testnet just replace the `currency.network` to `rinkeby`.
```
Comment (rittme) [inline packages/advanced-logic/specs/payment-network-erc20-proxy-contract-0.1.0-DRAFT.md]:
```suggestion
Note: to use the Rinkeby testnet, create a request with `currency.network` as `rinkeby`.
```

**Pull request merged**